File: go/internal/tools/twostage_destroy.go
package tools

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"slices"
	"strings"
	"time"

	"github.com/mark3labs/mcp-go/mcp"
	"google.golang.org/protobuf/proto"

	"github.com/chadit/LinodeMCP/go/internal/config"
	"github.com/chadit/LinodeMCP/go/internal/linode"
	"github.com/chadit/LinodeMCP/go/internal/twostage"
)

// twoStageSettings resolves the operator-tunable two-stage parameters from
// config into a twostage.Settings. It reads through resolveConfig so a
// hot-reloaded two_stage block takes effect on the next call. Non-positive TTL
// values are dropped here so the resolver falls back to the next level.
func twoStageSettings(cfg *config.Config) twostage.Settings {
	cfgTS := resolveConfig(cfg).TwoStage

	settings := twostage.Settings{OptIn: cfgTS.OptIn}

	if cfgTS.DefaultPlanTTLSeconds != nil && *cfgTS.DefaultPlanTTLSeconds > 0 {
		settings.DefaultTTL = time.Duration(*cfgTS.DefaultPlanTTLSeconds) * time.Second
	}

	if len(cfgTS.ToolTTLSeconds) > 0 {
		settings.ToolTTL = make(map[string]time.Duration, len(cfgTS.ToolTTLSeconds))

		for tool, secs := range cfgTS.ToolTTLSeconds {
			if secs > 0 {
				settings.ToolTTL[tool] = time.Duration(secs) * time.Second
			}
		}
	}

	return settings
}

// planResponse is the wire shape a mode:"plan" call returns: a richer dry-run
// that also carries a plan_id, an expiry, and a hash of the current state. A
// later mode:"apply" call re-fetches the state, re-hashes, and refuses if the
// hash moved.
type planResponse struct {
	PlanID           string        `json:"plan_id"`
	CreatedAt        string        `json:"created_at"`
	ExpiresAt        string        `json:"expires_at"`
	Tool             string        `json:"tool"`
	Environment      string        `json:"environment"`
	WouldExecute     DryRunRequest `json:"would_execute"`
	CurrentState     any           `json:"current_state"`
	CurrentStateHash string        `json:"current_state_hash"`

	// Dependency-walk enrichment, mirroring the dry-run response so a plan is
	// a strict superset of a dry-run. A tool with no DependencyWalk leaves
	// these unset (omitempty), keeping the shape stable.
	Dependencies []DryRunDependency  `json:"dependencies,omitempty"`
	SideEffects  []string            `json:"side_effects,omitempty"`
	BillingDelta *DryRunBillingDelta `json:"billing_delta,omitempty"`
	Warnings     []string            `json:"warnings,omitempty"`
}

// runTwoStageBranch handles the plan and apply branches for a destroy tool. It
// returns handled=false to let the caller fall through to the existing
// single-step flow whenever two-stage does not apply: a permitted yolo call (it
// dominates), a call without mode:"plan"/"apply", a server that did not attach
// a plan store (a direct unit-test call), or a tool that is not opted in.
func runTwoStageBranch(
	ctx context.Context,
	request *mcp.CallToolRequest,
	cfg *config.Config,
	action *DestructiveAction,
) (*mcp.CallToolResult, bool) {
	if yoloAllowedFromContext(ctx) {
		return nil, false
	}

	mode := request.GetString(paramMode, "")
	if mode != twostage.ModePlan && mode != twostage.ModeApply {
		return nil, false
	}

	store := PlanStoreFromContext(ctx)
	if store == nil {
		return nil, false
	}

	if !twoStageSettings(cfg).OptedIn(action.ToolName, action.capability()) {
		return nil, false
	}

	if mode == twostage.ModePlan {
		return runPlan(ctx, request, cfg, action, store), true
	}

	return runApply(ctx, request, cfg, action, store), true
}

// runPlan fetches the current state, hashes it, stores a single-use plan whose
// apply callback re-runs the destroy, and returns the plan preview.
func runPlan(
	ctx context.Context,
	request *mcp.CallToolRequest,
	cfg *config.Config,
	action *DestructiveAction,
	store *twostage.PlanStore,
) *mcp.CallToolResult {
	client, err := prepareClient(request, cfg)
	if err != nil {
		return mcp.NewToolResultError(err.Error())
	}

	state, fetchErr := action.FetchState(ctx, client)
	if fetchErr != nil {
		return mcp.NewToolResultError(fmt.Sprintf("Failed to fetch state for plan: %v", fetchErr))
	}

	hash, fields, hashErr := stateHashAndFields(state, action.HashIgnore)
	if hashErr != nil {
		return mcp.NewToolResultError(hashErr.Error())
	}

	planID, idErr := twostage.NewPlanID()
	if idErr != nil {
		return mcp.NewToolResultError(idErr.Error())
	}

	details, walkErr := planDependencies(ctx, client, action, state)
	if walkErr != nil {
		return mcp.NewToolResultError(fmt.Sprintf("Failed to compute plan dependencies: %v", walkErr))
	}

	now := time.Now()
	expires := now.Add(twoStageSettings(cfg).PlanTTL(action.ToolName))
	env := request.GetString(paramEnvironment, "")

	store.Put(&twostage.PlanEntry{
		ID:          planID,
		Tool:        action.ToolName,
		Environment: env,
		Args:        nonControlArgs(request.GetArguments()),
		StateHash:   hash,
		StateFields: fields,
		PlannedAt:   now,
		ExpiresAt:   expires,
		Apply: func(applyCtx context.Context) (*mcp.CallToolResult, error) {
			return executeDestroy(applyCtx, request, cfg, action)
		},
	})

	return buildPlanResponse(planID, now, expires, action, env, state, hash, &details)
}

// planDependencies runs the action's dependency walk (if any) so the plan
// response is a strict superset of a dry-run. A nil walk yields empty details.
func planDependencies(
	ctx context.Context,
	client *linode.Client,
	action *DestructiveAction,
	state any,
) (DryRunDetails, error) {
	if action.DependencyWalk == nil {
		return DryRunDetails{}, nil
	}

	return action.DependencyWalk(ctx, client, state)
}

// runApply classifies the referenced plan, resolves the branch, and either runs
// the stored apply callback (after a drift check) or returns a refusal.
func runApply(
	ctx context.Context,
	request *mcp.CallToolRequest,
	cfg *config.Config,
	action *DestructiveAction,
	store *twostage.PlanStore,
) *mcp.CallToolResult {
	planID := request.GetString(paramPlanID, "")

	lookup, entry, changedFields, errResult := classifyPlan(ctx, request, cfg, action, store, planID)
	if errResult != nil {
		return errResult
	}

	decision := twostage.Resolve(twostage.Request{
		Capability:      action.capability(),
		TwoStageOptedIn: true,
		Mode:            twostage.ModeApply,
		PlanID:          planID,
		PlanLookup:      lookup,
	})
	if decision.Branch != twostage.BranchApply {
		return refusalResult(decision.ErrCode, planID, changedFields)
	}

	result, applyErr := entry.Apply(ctx)
	if applyErr != nil {
		return mcp.NewToolResultError(applyErr.Error())
	}

	store.Remove(planID)

	return result
}

// classifyPlan looks up the plan, checks the supplied args against the stored
// args, and re-fetches plus re-hashes the state to detect drift. The returned
// PlanLookup feeds twostage.Resolve. A non-nil result is a hard error (client
// or fetch failure) that short-circuits before resolution.
func classifyPlan(
	ctx context.Context,
	request *mcp.CallToolRequest,
	cfg *config.Config,
	action *DestructiveAction,
	store *twostage.PlanStore,
	planID string,
) (twostage.PlanLookup, *twostage.PlanEntry, []string, *mcp.CallToolResult) {
	entry, lookupErr := store.Get(planID)
	if errors.Is(lookupErr, twostage.ErrPlanExpired) {
		return twostage.PlanLookupExpired, nil, nil, nil
	}

	if lookupErr != nil {
		return twostage.PlanLookupUnknown, nil, nil, nil
	}

	supplied := nonControlArgs(request.GetArguments())
	if len(supplied) > 0 && !argsEqual(supplied, entry.Args) {
		return twostage.PlanLookupArgsMismatch, entry, nil, nil
	}

	client, clientErr := prepareClient(request, cfg)
	if clientErr != nil {
		return twostage.PlanLookupNotApplicable, nil, nil, mcp.NewToolResultError(clientErr.Error())
	}

	state, fetchErr := action.FetchState(ctx, client)
	if fetchErr != nil {
		return twostage.PlanLookupNotApplicable, nil, nil,
			mcp.NewToolResultError(fmt.Sprintf("Failed to re-fetch state for apply: %v", fetchErr))
	}

	hash, fields, hashErr := stateHashAndFields(state, action.HashIgnore)
	if hashErr != nil {
		return twostage.PlanLookupNotApplicable, nil, nil, mcp.NewToolResultError(hashErr.Error())
	}

	if hash != entry.StateHash {
		return twostage.PlanLookupDrifted, entry, changedFieldNames(entry.StateFields, fields), nil
	}

	return twostage.PlanLookupValid, entry, nil, nil
}

// executeDestroy runs the real delete: prepare the client, execute, and marshal
// the success body. It backs both the single-step path in RunDestructiveAction
// and the apply callback a plan stores.
func executeDestroy(
	ctx context.Context,
	request *mcp.CallToolRequest,
	cfg *config.Config,
	action *DestructiveAction,
) (*mcp.CallToolResult, error) {
	client, err := prepareClient(request, cfg)
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}

	if execErr := action.Execute(ctx, client); execErr != nil {
		return mcp.NewToolResultError(fmt.Sprintf("%s failed: %v", action.ToolName, execErr)), nil
	}

	return marshalDestroySuccess(action.Success())
}

// marshalDestroySuccess serializes a destroy Success body. A proto.Message goes
// through the proto-canonical marshaller so its output matches the Python side;
// everything else falls back to plain JSON for the legacy id-echo maps.
func marshalDestroySuccess(success any) (*mcp.CallToolResult, error) {
	if msg, ok := success.(proto.Message); ok {
		return MarshalProtoToolResponse(msg)
	}

	return MarshalToolResponse(success)
}

// stateHash returns a stable hash of the resource state with the named cosmetic
// fields stripped first, so a plan does not refuse on drift the user never
// caused. Go's json.Marshal sorts map keys, so the same state always encodes
// the same way.
// stateHashAndFields hashes a resource's state for drift detection and also
// returns its normalized top-level field map with the hash-ignore fields
// stripped. The map lets the apply path report which fields changed on a drift
// refusal. It is nil when the state does not serialize to a JSON object (for
// example a bare array), in which case the whole payload is hashed and no
// per-field diff is available. Plan and apply both call this, so the hash is
// computed identically on both sides.
func stateHashAndFields(state any, ignore []string) (string, map[string]any, error) {
	data, err := json.Marshal(state)
	if err != nil {
		return "", nil, fmt.Errorf("marshal state for hash: %w", err)
	}

	trimmed := bytes.TrimSpace(data)
	if len(trimmed) == 0 || trimmed[0] != '{' {
		sum := sha256.Sum256(data)

		return "sha256:" + hex.EncodeToString(sum[:]), nil, nil
	}

	var obj map[string]any
	if err := json.Unmarshal(data, &obj); err != nil {
		return "", nil, fmt.Errorf("unmarshal state for hash: %w", err)
	}

	for _, field := range ignore {
		delete(obj, field)
	}

	stripped, err := json.Marshal(obj)
	if err != nil {
		return "", nil, fmt.Errorf("re-marshal stripped state for hash: %w", err)
	}

	sum := sha256.Sum256(stripped)

	return "sha256:" + hex.EncodeToString(sum[:]), obj, nil
}

// changedFieldNames returns the sorted top-level keys whose values differ
// between the planned and current field maps (added, removed, or changed). It
// drives the changed-fields list in a drift refusal. Comparison is by canonical
// JSON encoding, so key ordering within nested objects is irrelevant. Returns
// nil when either map is nil (no per-field diff available).
func changedFieldNames(planned, current map[string]any) []string {
	if planned == nil || current == nil {
		return nil
	}

	seen := make(map[string]struct{}, len(planned)+len(current))
	for key := range planned {
		seen[key] = struct{}{}
	}

	for key := range current {
		seen[key] = struct{}{}
	}

	var changed []string

	for key := range seen {
		if !jsonEqual(planned[key], current[key]) {
			changed = append(changed, key)
		}
	}

	slices.Sort(changed)

	return changed
}

// jsonEqual reports whether two values have the same canonical JSON encoding.
func jsonEqual(left, right any) bool {
	leftBytes, leftErr := json.Marshal(left)
	rightBytes, rightErr := json.Marshal(right)

	if leftErr != nil || rightErr != nil {
		return false
	}

	return bytes.Equal(leftBytes, rightBytes)
}

// buildPlanResponse renders the plan preview a mode:"plan" call returns.
func buildPlanResponse(
	planID string,
	createdAt, expiresAt time.Time,
	action *DestructiveAction,
	environment string,
	state any,
	hash string,
	details *DryRunDetails,
) *mcp.CallToolResult {
	result, err := MarshalToolResponse(planResponse{
		PlanID:           planID,
		CreatedAt:        createdAt.UTC().Format(time.RFC3339),
		ExpiresAt:        expiresAt.UTC().Format(time.RFC3339),
		Tool:             action.ToolName,
		Environment:      environment,
		WouldExecute:     DryRunRequest{Method: action.Method, Path: action.Path},
		CurrentState:     state,
		CurrentStateHash: hash,
		Dependencies:     details.Dependencies,
		SideEffects:      details.SideEffects,
		BillingDelta:     details.BillingDelta,
		Warnings:         details.Warnings,
	})
	if err != nil {
		return mcp.NewToolResultError(err.Error())
	}

	return result
}

// refusalResult maps an apply refusal code to a message that tells the model
// how to recover.
func refusalResult(errCode, planID string, changedFields []string) *mcp.CallToolResult {
	switch errCode {
	case twostage.ErrCodePlanExpired:
		return mcp.NewToolResultError(fmt.Sprintf(
			"PLAN_EXPIRED: plan %q has expired. Create a new plan with mode: \"plan\".", planID,
		))
	case twostage.ErrCodePlanArgsMismatch:
		return mcp.NewToolResultError(fmt.Sprintf(
			"PLAN_ARGS_MISMATCH: args supplied at apply time differ from plan %q. "+
				"Apply without passing args (the plan retains them), or create a new plan.", planID,
		))
	case twostage.ErrCodePlanDrift:
		changed := "one or more fields"
		if len(changedFields) > 0 {
			changed = strings.Join(changedFields, ", ")
		}

		return mcp.NewToolResultError(fmt.Sprintf(
			"PLAN_DRIFT_DETECTED: the resource changed since plan %q was created "+
				"(changed fields: %s). Create a new plan with mode: \"plan\" and review "+
				"before applying.", planID, changed,
		))
	default:
		return mcp.NewToolResultError(fmt.Sprintf(
			"PLAN_NOT_FOUND: no plan with id %q. Create a new plan with mode: \"plan\". "+
				"Plans do not persist across a server restart.", planID,
		))
	}
}

// nonControlArgs returns the tool's own arguments with the two-stage and
// confirmation control flags stripped, so the apply-time args comparison
// matches only the underlying call.
func nonControlArgs(args map[string]any) map[string]any {
	control := map[string]struct{}{
		paramMode:                {},
		paramPlanID:              {},
		paramConfirm:             {},
		paramDryRun:              {},
		paramConfirmedDryRun:     {},
		paramConfirmBypassDryRun: {},
		paramYolo:                {},
	}

	out := make(map[string]any, len(args))

	for key, val := range args {
		if _, isControl := control[key]; !isControl {
			out[key] = val
		}
	}

	return out
}

// argsEqual reports whether two argument maps are deep-equal. JSON-decoded
// values compare by value, so this implements the spec's strict-equality check.
func argsEqual(supplied, stored map[string]any) bool {
	if len(supplied) != len(stored) {
		return false
	}

	suppliedJSON, errA := json.Marshal(supplied)
	storedJSON, errB := json.Marshal(stored)

	if errA != nil || errB != nil {
		return false
	}

	return bytes.Equal(suppliedJSON, storedJSON)
}

File: python/src/linodemcp/tools/twostage_destroy.py
"""Two-stage plan/apply flow for destructive Python tools.

Mirrors ``go/internal/tools/twostage_destroy.go``. ``run_two_stage_destroy``
reads the plan store the server published on the ContextVar, handles
``mode:"plan"`` and ``mode:"apply"``, and returns ``None`` for everything else
so the caller falls through to its existing single-step path.
"""

from __future__ import annotations

import dataclasses
import hashlib
import json
from datetime import UTC, datetime, timedelta
from typing import TYPE_CHECKING, Any, cast

import httpx
from mcp.types import TextContent

from linodemcp import twostage
from linodemcp.config import EnvironmentNotFoundError
from linodemcp.genpb.linode.mcp.v1 import dryrun_pb2
from linodemcp.linode import APIError, NetworkError
from linodemcp.profiles import Capability
from linodemcp.tools import helpers
from linodemcp.tools.proto_response import serialize_preview_envelope
from linodemcp.twostage.store import PlanEntry, PlanExpiredError, PlanNotFoundError

if TYPE_CHECKING:
    from collections.abc import Awaitable, Callable

    from linodemcp.config import Config
    from linodemcp.linode import RetryableClient
    from linodemcp.twostage.store import PlanStore

# Control flags stripped before comparing apply-time args to the stored args.
_CONTROL_KEYS = frozenset(
    {
        "mode",
        "plan_id",
        "confirm",
        "dry_run",
        "confirmed_dry_run",
        "confirm_bypass_dry_run",
        "yolo",
    }
)

# Errors a state fetch can raise, matching execute_tool's handling.
_FETCH_ERRORS = (
    EnvironmentNotFoundError,
    ValueError,
    APIError,
    NetworkError,
    httpx.HTTPError,
)

type _FetchState = Callable[[RetryableClient], Awaitable[Any]]
type _Execute = Callable[[RetryableClient], Awaitable[dict[str, Any]]]
# Per-tool dependency walk run at plan time. Mirrors the dry-run details_fn:
# given a live client and the fetched state, it returns the same DryRunDetails
# (dependencies / side_effects / billing_delta / warnings) the dry-run path
# surfaces, so a plan reads like a dry-run preview.
type _DependencyWalk = Callable[
    [RetryableClient, Any], Awaitable[helpers.DryRunDetails]
]


def _json_default(obj: object) -> Any:
    if dataclasses.is_dataclass(obj) and not isinstance(obj, type):
        # Same keyword-escape stripping as the dry-run path, so a plan's
        # current_state carries the wire field names ("in", not "in_").
        return dataclasses.asdict(obj, dict_factory=helpers.keyword_escape_dict_factory)
    return str(obj)


def _state_hash_and_fields(
    state: Any, hash_ignore: list[str]
) -> tuple[str, dict[str, Any] | None]:
    """Hash the state for drift detection and return its normalized top-level
    field map with hash-ignore fields stripped. The map lets the apply path
    name the changed fields on a drift refusal; it is None when the state does
    not serialize to a JSON object. Plan and apply both call this, so the hash
    is identical on both sides. Mirrors the Go stateHashAndFields.
    """
    serialized = json.dumps(state, sort_keys=True, default=_json_default)
    parsed = json.loads(serialized)
    if not isinstance(parsed, dict):
        return "sha256:" + hashlib.sha256(serialized.encode()).hexdigest(), None

    obj = cast("dict[str, Any]", parsed)
    for field in hash_ignore:
        obj.pop(field, None)

    stripped = json.dumps(obj, sort_keys=True)
    return "sha256:" + hashlib.sha256(stripped.encode()).hexdigest(), obj


def _changed_field_names(
    planned: dict[str, Any] | None, current: dict[str, Any] | None
) -> list[str]:
    """Return the sorted top-level keys whose values differ between the planned
    and current field maps. Drives the changed-fields list in a drift refusal.
    Returns an empty list when either map is None (no per-field diff available).
    """
    if planned is None or current is None:
        return []

    def encode(value: Any) -> str:
        return json.dumps(value, sort_keys=True, default=_json_default)

    keys = set(planned) | set(current)
    changed = [
        key for key in keys if encode(planned.get(key)) != encode(current.get(key))
    ]
    return sorted(changed)


def _non_control_args(arguments: dict[str, Any]) -> dict[str, Any]:
    return {key: value for key, value in arguments.items() if key not in _CONTROL_KEYS}


def _text(message: str) -> list[TextContent]:
    return [TextContent(type="text", text=message)]


def _refusal(
    err_code: str, plan_id: str, changed_fields: list[str] | None = None
) -> list[TextContent]:
    changed = ", ".join(changed_fields) if changed_fields else "one or more fields"
    messages = {
        twostage.ERR_PLAN_EXPIRED: (
            f"PLAN_EXPIRED: plan {plan_id!r} has expired. "
            'Create a new plan with mode: "plan".'
        ),
        twostage.ERR_PLAN_ARGS_MISMATCH: (
            f"PLAN_ARGS_MISMATCH: args supplied at apply time differ from plan "
            f"{plan_id!r}. Apply without passing args, or create a new plan."
        ),
        twostage.ERR_PLAN_DRIFT: (
            f"PLAN_DRIFT_DETECTED: the resource changed since plan {plan_id!r} "
            f"was created (changed fields: {changed}). "
            'Create a new plan with mode: "plan" and review first.'
        ),
    }
    default = (
        f"PLAN_NOT_FOUND: no plan with id {plan_id!r}. "
        'Create a new plan with mode: "plan". Plans do not persist across a restart.'
    )
    return _text(messages.get(err_code, default))


async def run_two_stage_destroy(
    cfg: Config,
    arguments: dict[str, Any],
    *,
    tool_name: str,
    method: str,
    path: str,
    fetch_state: _FetchState,
    execute: _Execute,
    hash_ignore: list[str] | None = None,
    dependency_walk: _DependencyWalk | None = None,
    capability: Capability = Capability.Destroy,
) -> list[TextContent] | None:
    """Handle plan/apply for a destroy tool, or None to fall through.

    Returns None when two-stage does not apply: a yolo call (it dominates via
    the server's single-step path), no plan store on the context, a call
    without mode:"plan"/"apply", or a tool that is not opted in.

    capability is the tool's profile capability, consulted at the opt-in gate.
    It defaults to Destroy (every delete tool), so a CapWrite tool like
    instance_resize passes Capability.Write to stay opt-in by config only.
    """
    if arguments.get("yolo") is True:
        return None

    store = twostage.plan_store_from_context()
    if store is None:
        return None

    mode = arguments.get("mode", "")
    if mode not in (twostage.MODE_PLAN, twostage.MODE_APPLY):
        return None

    settings = _two_stage_settings(cfg)
    if not settings.opted_in(tool_name, capability):
        return None

    ignore = hash_ignore or []
    if mode == twostage.MODE_PLAN:
        return await _run_plan(
            cfg,
            arguments,
            tool_name,
            method,
            path,
            fetch_state,
            execute,
            store,
            ignore,
            settings,
            dependency_walk,
        )

    return await _run_apply(cfg, arguments, fetch_state, store, ignore, capability)


def _two_stage_settings(cfg: Config) -> twostage.Settings:
    """Resolve the operator-tunable two-stage parameters from config.

    Non-positive TTL values are dropped here so the resolver falls back to the
    next level (per-tool to default to built-in). Mirrors the Go
    twoStageSettings helper.
    """
    two_stage = cfg.two_stage

    default_ttl: timedelta | None = None
    if (
        two_stage.default_plan_ttl_seconds is not None
        and two_stage.default_plan_ttl_seconds > 0
    ):
        default_ttl = timedelta(seconds=two_stage.default_plan_ttl_seconds)

    tool_ttl = {
        tool: timedelta(seconds=secs)
        for tool, secs in two_stage.tool_ttl_seconds.items()
        if secs > 0
    }

    return twostage.Settings(
        default_ttl=default_ttl, tool_ttl=tool_ttl, opt_in=dict(two_stage.opt_in)
    )


async def _run_plan(
    cfg: Config,
    arguments: dict[str, Any],
    tool_name: str,
    method: str,
    path: str,
    fetch_state: _FetchState,
    execute: _Execute,
    store: PlanStore,
    hash_ignore: list[str],
    settings: twostage.Settings,
    dependency_walk: _DependencyWalk | None,
) -> list[TextContent]:
    # Fetch the state and run the dependency walk under one client so the plan
    # body reads like a dry-run preview (dependencies / side_effects /
    # billing_delta / warnings). Mirrors the Go runPlan + planDependencies.
    async def _fetch_and_walk(
        client: RetryableClient,
    ) -> tuple[Any, helpers.DryRunDetails]:
        state = await fetch_state(client)
        details: helpers.DryRunDetails = {}
        if dependency_walk is not None:
            details = await dependency_walk(client, state)
        return state, details

    try:
        state, details = await helpers.with_client(cfg, arguments, _fetch_and_walk)
    except _FETCH_ERRORS as exc:
        return _text(f"Failed to fetch state for plan: {exc}")

    state_hash, state_fields = _state_hash_and_fields(state, hash_ignore)
    plan_id = twostage.new_plan_id()
    now = datetime.now(UTC)
    expires = now + settings.plan_ttl(tool_name)
    environment = arguments.get("environment", "")

    async def _apply() -> list[TextContent]:
        return await helpers.execute_tool(cfg, arguments, f"apply {tool_name}", execute)

    await store.put(
        PlanEntry(
            id=plan_id,
            tool=tool_name,
            environment=environment,
            args=_non_control_args(arguments),
            state_hash=state_hash,
            planned_at=now,
            expires_at=expires,
            apply=_apply,
            state_fields=state_fields,
        )
    )

    raw: dict[str, Any] = {
        "plan_id": plan_id,
        "created_at": _rfc3339(now),
        "expires_at": _rfc3339(expires),
        "tool": tool_name,
        "environment": environment,
        "would_execute": {"method": method, "path": path},
        "current_state": state,
        "current_state_hash": state_hash,
        **details,
    }
    # Round-trip through JSON first so a dataclass current_state collapses to the
    # plain dict ParseDict accepts, mirroring the Go builder. The PlanResponse
    # proto then makes the envelope proto-canonical on both languages.
    plain = cast("dict[str, Any]", json.loads(json.dumps(raw, default=_json_default)))
    result = serialize_preview_envelope(plain, dryrun_pb2.PlanResponse())
    return [TextContent(type="text", text=json.dumps(result, indent=2))]


def _rfc3339(moment: datetime) -> str:
    """Format a UTC datetime as RFC3339 with a Z suffix and whole seconds.

    Matches Go's time.RFC3339 output (no fractional seconds, Z zone) so the plan
    timestamps read identically on both languages, replacing Python's
    isoformat() which emitted a +00:00 offset with microseconds.
    """
    return (
        moment.astimezone(UTC).replace(microsecond=0).isoformat().replace("+00:00", "Z")
    )


async def _run_apply(
    cfg: Config,
    arguments: dict[str, Any],
    fetch_state: _FetchState,
    store: PlanStore,
    hash_ignore: list[str],
    capability: Capability,
) -> list[TextContent]:
    plan_id = arguments.get("plan_id", "")
    lookup, entry, changed_fields, error = await _classify_plan(
        cfg, arguments, fetch_state, store, plan_id, hash_ignore
    )
    if error is not None:
        return error

    decision = twostage.resolve(
        twostage.Request(
            capability=capability,
            two_stage_opted_in=True,
            mode=twostage.MODE_APPLY,
            plan_id=plan_id,
            plan_lookup=lookup,
        )
    )
    if decision.branch != twostage.Branch.APPLY or entry is None:
        return _refusal(decision.err_code, plan_id, changed_fields)

    result = await entry.apply()
    await store.remove(plan_id)
    return result


async def _classify_plan(
    cfg: Config,
    arguments: dict[str, Any],
    fetch_state: _FetchState,
    store: PlanStore,
    plan_id: str,
    hash_ignore: list[str],
) -> tuple[twostage.PlanLookup, PlanEntry | None, list[str], list[TextContent] | None]:
    # The third element is the list of changed top-level field names, populated
    # only on a drift verdict so the refusal can name them. Mirrors the Go
    # classifyPlan, whose fourth return is the changed-field slice.
    try:
        entry = await store.get(plan_id)
    except PlanNotFoundError:
        return twostage.PlanLookup.UNKNOWN, None, [], None
    except PlanExpiredError:
        return twostage.PlanLookup.EXPIRED, None, [], None

    supplied = _non_control_args(arguments)
    if supplied and supplied != entry.args:
        return twostage.PlanLookup.ARGS_MISMATCH, entry, [], None

    try:
        state = await helpers.with_client(cfg, arguments, fetch_state)
    except _FETCH_ERRORS as exc:
        return (
            twostage.PlanLookup.NOT_APPLICABLE,
            None,
            [],
            _text(f"Failed to re-fetch state for apply: {exc}"),
        )

    current_hash, current_fields = _state_hash_and_fields(state, hash_ignore)
    if current_hash != entry.state_hash:
        changed = _changed_field_names(entry.state_fields, current_fields)
        return twostage.PlanLookup.DRIFTED, entry, changed, None

    return twostage.PlanLookup.VALID, entry, [], None
